### ISSUE TITLE
Nupkg deployments to Azure Web Apps fail the release package part of the deployment (Error - The file 'C:\Octopus\Tentacle\Files\packagename.zip' already exists).

### ISSUE THREAD

**Issue by Clare-Octopus** [OctopusDeploy/Issues#8727]
### Severity

Not blocking customers, the package itself is deployed, its just the release part of the deployment fails.

### Version

2024.2.3973, 2024.1.12087

### Latest Version

I could reproduce the problem in the latest build

### What happened?

When trying to deploy a nupkg to an Azure App Service using the Azure App Service step in a deployment, the step will convert the nupkg to a .zip file (as the step itself is a zip deploy step). The package itself will get sent to the target Web App so the deployment itself will succeed and your application will be deployed, but on the release packages stage you will see one of three errors depending on what workers you use for the step:

Cloud Hosted Windows Worker Pool (our dynamic worker):

`The process cannot access the file 'C:\Octopus\Tentacle\Files\packagename.zip' because it is being used by another process.`

On retry of the deployment you will see:

`The file 'C:\Octopus\Tentacle\Files\packagename.zip' already exists.`

During my testing I found if using a normal windows worker or one of our Ubuntu Dynamic Cloud workers you will see:

`Response status code does not indicate success: 401 (Unauthorized).`

Your applications will be deployed but the Octopus deployment itself will show as a failure.

### Reproduction

1. Ensure you have a WebApp to deploy to in Azure.
2. Set that WebApp up in Octopus as a target.
3. Ensure you have a basic .nupkg in the Octopus Built-In feed or an external feed.
4. Create a basic project with a `Deploy an App Service` step.
5. Create a release and deploy the nupkg file to that webapp.
6. See the errors.

### Error and Stacktrace

```shell
08:12:10   Verbose  |       Target Site: ClareWebApp
08:12:10   Verbose  |       Publishing C:\Octopus\Tentacle\Files\RandomQuotesTCNuget@S2.0.1@8E53B723E7435A46A4E14257A6A2A940.zip to https://clarewebapp.scm.azurewebsites.net:443/api/zipdeploy
08:14:20   Error    |       Running rollback behaviours...
08:14:20   Error    |       The process cannot access the file 'C:\Octopus\Tentacle\Files\RandomQuotesTCNuget@S2.0.1@8E53B723E7435A46A4E14257A6A2A940.zip' because it is being used by another process.
08:14:20   Error    |       System.IO.IOException
08:14:20   Error    |       at System.IO.FileSystem.DeleteFile(String fullPath)
08:14:20   Error    |       at System.IO.File.Delete(String path)
08:14:20   Error    |       at Calamari.AzureAppService.Behaviors.AzureAppServiceZipDeployBehaviour.CleanupUploadFile(String uploadPath) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.AzureAppService\Behaviors\AzureAppServiceZipDeployBehaviour.cs:line 358
08:14:20   Error    |       at Calamari.AzureAppService.Behaviors.AzureAppServiceZipDeployBehaviour.Execute(RunningDeployment context) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.AzureAppService\Behaviors\AzureAppServiceZipDeployBehaviour.cs:line 176
08:14:20   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.ExecuteBehaviour(RunningDeployment context, IBehaviour behaviour) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.Common\Plumbing\Pipeline\PipelineCommand.cs:line 165
08:14:20   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.Execute(ILifetimeScope lifetimeScope, IVariables variables) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.Common\Plumbing\Pipeline\PipelineCommand.cs:line 61
08:14:20   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.Execute(ILifetimeScope lifetimeScope, IVariables variables) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.Common\Plumbing\Pipeline\PipelineCommand.cs:line 85
08:14:20   Error    |       at Calamari.Common.CalamariFlavourProgramAsync.Run(String[] args) in C:\BuildAgent\work\e0cefbed4ad11812\source\Calamari.Common\CalamariFlavourProgramAsync.cs:line 144
08:14:20   Verbose  |       Process C:\Windows\system32\WindowsPowershell\v1.0\PowerShell.exe in C:\Octopus\Tentacle\Work\EzhrA7ICpkeJBgDV3Ji2A exited with code 100
08:14:21   Verbose  |       Updating manifest with output variables
08:14:21   Verbose  |       Updating manifest with action evaluated variables
08:14:21   Verbose  |       Updating manifest with output variables
08:14:21   Verbose  |       Updating manifest with action evaluated variables
08:14:21   Fatal    |       The remote script failed with exit code 100
08:14:21   Fatal    |       The action Deploy an Azure App Service on ClareWebApp failed
```

### More Information

- Initial ticket raised (internal) - https://octopus.zendesk.com/agent/tickets/178242

- Slack Discussion (internal) - https://octopusdeploy.slack.com/archives/C01Q95KPRM4/p1712244924367929

- Reproduction (internal) - https://octopus-operations.octopus.app/app#/Spaces-422/projects/ticket-178242-test/deployments?groupBy=Channel

Engineers Note, my WebApp wont be around for long so you may need to create a new one if using my reproduction for testing.

### Workaround

Configure one of these settings in the Azure App Service step you are using:

- Substitute Variables In Templates
- Structured Configuration Variables
- .NET Configuration Transforms

I personally would try adding a filename into the `Structured Configuration Variables` section of the step as you dont have to configure any additional parameters in that feature and you can save the step without it impacting the deployment.

**Use a filename you know is not going to be in your app package though, for example temporary.json.**

![image](https://github.com/OctopusDeploy/Issues/assets/92790462/b453bdfc-1bef-41d0-ba23-d334a7734748)

**Comment by Clare-Octopus** [comment]
Two other users seeing this issue (internal links):

https://octopus.zendesk.com/agent/tickets/183291
https://octopus.zendesk.com/agent/tickets/185329

**Comment by patriksvensson** [comment]
Any updates on this issue? Although the publish to Azure succeeds, the Octopus task fails, which prevents me from deploying to the next environment in the lifecycle.

**Comment by patriksvensson** [comment]
Sorry, didn't see the notice above about being able to edit the state.

**Comment by patriksvensson** [comment]
The deployment was working before, but now it's stopped working, so we would really need some kind of workaround for this.

**Comment by Clare-Octopus** [comment]
Hey @patriksvensson thank you for posting on here about your issues. You can see there is a fix in place for this in the form of a retry clause in the PR linked in this ticket - https://github.com/OctopusDeploy/Calamari/pull/1250.

You are unable to see the merge, though, as it's internal. This was merged into a Calamari version specific to an Octopus version. The Calamari version this was updated in was `27.1.14` and above which is bundled with Octopus versions `2024.2.8080` and above. 

That Octopus version is currently only available to our Cloud customers, most of whom are now running a 2024.3 version. I can see your previous ticket with us, which I worked through with you, and your Cloud instance is on `2024.3.531`. 

So your instance should have the retry fix in place and the deployment should retry the deletion of that file. If it is not doing that could you please either respond to your previous email thread you had in with us or email us again at support@octopus.com and quote ticket ref `178242` and we will continue to investigate this with you.

We have still not been able to replicate this ourselves except the one time I managed to do it but have not been able to do it since so this is very very intermittent and since we have not managed to do it a second time in testing its been difficult to get an actual fix for it unfortunately.

Hopefully we can continue to help over email, I look forward to continuing to help.
Kind Regards,
Clare

**Comment by Clare-Octopus** [comment]
Good morning all,

Our engineers have finally managed to replicate this and have put a PR in place for the fix, which you can view [here](https://github.com/OctopusDeploy/Calamari/pull/1275). They will backport the fix to a 2024.1 version for our Octopus Server customers and include it with our regular Octopus Cloud upgrade cadence. Please do keep subscribed to this GitHub issue to ensure you get the email when the fix has been put in place and in which Octopus versions.

I have updated the GitHub issue with a workaround that the engineers mentioned works in their testing. However, I cannot test it since I personally still cannot replicate the issue. Please do reach out via email (support@octopus.com) if enabling any of these within the App Deploy step does not work for you.

`The workaround for this issue until we can get the fix out is:`

To configure one of these settings in the Azure App Service step you are using:

- Substitute Variables In Templates
- Structured Configuration Variables
- .NET Configuration Transforms

I personally would try adding a filename into the `Structured Configuration Variables` section of the step as you dont have to configure any additional parameters in that feature and you can save the step without it impacting the deployment. 

**Use a filename you know is not going to be in your app package though, for example `temporary.json`.**

![image](https://github.com/OctopusDeploy/Issues/assets/92790462/58946372-0eec-4b74-bdad-daf5f8cedcbf)

Hopefully, we can get the permanent fix out soon but please reach out via email if the workaround does not work for you.

Kind Regards,
Clare

**Comment by patriksvensson** [comment]
@Clare-Octopus Oh wow, thanks! Will try it out immediately! Thanks!

**Comment by patriksvensson** [comment]
@Clare-Octopus The deploy is still failing for me, even with the workaround in place. Created a new deployment after adding the variable substitution workaround.

![CleanShot 2024-06-02 at 13 35 46@2x](https://github.com/OctopusDeploy/Issues/assets/357872/1dbd7266-3951-4e31-8a09-6b15bcfdb135)

**Comment by Clare-Octopus** [comment]
Hey @patriksvensson,

Thank you for confirming that the workaround does not work for you. That is disappointing news! Our engineer is unsure why since, according to our code on that step, the cleanup of the zip file (which is failing the deployment) is not triggered when using one of those configuration file options in the workaround.

However, our engineer has another fix in place, which we hope resolves this issue for good for our customers. 

For your specific use case, since you are the only customer we know of seeing this on Octopus Cloud, the fix has been merged into `2024.3.2146`, which your instance is **_not currently_** running. I am hoping you will be upgraded to that version or higher at some point next week, as that version has not been deployed to our testing environments yet so I am unable to put you on it. It is usually better to wait for your instance to auto-upgrade so that it has passed our testing phases.

I will keep an eye out next week to see if you have been upgraded and will let you know, but you can also do the same and take a look at your instances' version every now and then next week. If you notice it goes to `2024.3.2146` or above, you can test the step and let us know if it still fails or is now successful.

For our **Octopus Server customers** experiencing this, we are looking to backport that fix into a `2024.2` version, which we hope will be available to download over the next few weeks. This GitHub issue will be updated with the versions the fix is in so all customers know.

Kind Regards,
Clare

**Comment by patriksvensson** [comment]
@Clare-Octopus I just wanted to inform you that the deployment worked now! Thank you for all the help with this issue. A+ support!

**Comment by Clare-Octopus** [comment]
Hey @patriksvensson,

Amazing news this is now working for you, thank you for reaching out to confirm that is the case as we can now close the issue off on our end. Also thank you for your kind words on our Support, we do not like bugs so we do try our best to squash them when we do find any, this one was hard to replicate so thank you for your patience whilst we worked through getting a fix out for this. 

If anyone else is experiencing this issue after an upgrade to 2024.3.2146 or above on Octopus Cloud please do let us know via our email at support@octopus.com. I am still waiting on what Octopus Server version this will be included so will get our engineers to close this issue off with the versions the fix is in so everyone can download it and upgrade when they can.

Kind Regards,
Clare

**Comment by octoreleasebot** [comment]
Release Note: Fixed an issue where async zip deploy failed to clean up packages resulting in failed tasks.

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2024.1 | 2024.1.12875 |
| 2024.2 | _tbc_ |
| 2024.3 | _tbc_ |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2024.1 | 2024.1.12875 |
| 2024.2 | _tbc_ |
| 2024.3 | 2024.3.4044 |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2024.1 | 2024.1.12875 |
| 2024.2 | 2024.2.9235 |
| 2024.3 | 2024.3.4044 |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2024.1 | 2024.1.12875 |
| 2024.2 | 2024.2.9235 |
| 2024.3 | 2024.3.4044 |
| 2024.4+ | _all releases_ |
